[Core][Scheme] Correcting & unifying implementation for InitializeNonLinearIteration

philbucher opened the pull request:
`scheme.h` calls the corresponding methods of elems/conds/constaints of `InitializeSolutionStep`, `FinalizeSolutionStep` and `FinalizeNonLinIteration`
The corresponding implementation for `InitializeNonLinIteration` was missing and is added in this PR
I have not run the tests yet since I want to first discuss if this is the correct change (I would say yes)

found by @KlausBSautter

philbucher committed:
adding implementation for InitializeNonLinearIteration

loumalouomega reviewed (dismissed):
I think I fixed this in some derived schemes, but I forgot to add on the base one. What about MPC?

philbucher commented on kratos/solving_strategies/schemes/scheme.h:
```diff
@@ -502,6 +502,39 @@ class Scheme
+
+        // Initialize non-linear iteration for all of the constraints
+        #pragma omp parallel for
+        for(int i=0; i<static_cast<int>(rModelPart.MasterSlaveConstraints().size()); ++i) {
```
@loumalouomega

RiccardoRossi replied:
i think that this was deliberately not called here but rather directly within the builder and solver.

philbucher replied:
nope, the BuilderAndSolver never did this, I checked
Actually none did this I am quite sure

loumalouomega commented:
OK, I haven't seen it. Go on with the merge ⛜

El lun., 24 jun. 2019 13:14, Philipp Bucher <notifications@github.com>
escribió:

> *@philbucher* commented on this pull request.
> ------------------------------
>
> In kratos/solving_strategies/schemes/scheme.h
> <https://github.com/KratosMultiphysics/Kratos/pull/5121#discussion_r296691658>
> :
>
> > +        // Definition of the first condition iterator
> +        const auto it_cond_begin = rModelPart.ConditionsBegin();
> +
> +        // Initialize non-linear iteration for all of the conditions
> +        #pragma omp parallel for
> +        for(int i=0; i<static_cast<int>(rModelPart.Conditions().size()); ++i) {
> +            auto it_cond = it_cond_begin + i;
> +            it_cond->InitializeNonLinearIteration(r_current_process_info);
> +        }
> +
> +        // Definition of the first constraint iterator
> +        const auto it_const_begin = rModelPart.MasterSlaveConstraintsBegin();
> +
> +        // Initialize non-linear iteration for all of the constraints
> +        #pragma omp parallel for
> +        for(int i=0; i<static_cast<int>(rModelPart.MasterSlaveConstraints().size()); ++i) {
>
> @loumalouomega <https://github.com/loumalouomega>
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/KratosMultiphysics/Kratos/pull/5121?email_source=notifications&email_token=AEYQZADQV2LZ4PJV757Y34LP4C3ANA5CNFSM4H24PQA2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOB4NCODA#pullrequestreview-253372172>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEYQZADOFWHK3FIPJQQGD2DP4C3ANANCNFSM4H24PQAQ>
> .
>

loumalouomega commented:
What is missing to merge?

philbucher commented:
> What is missing to merge?

Testing it more properly (aka running the tests)
At the time I did this I had no time and I wanted to put it up for discussion first

RiccardoRossi reviewed (changes requested):
this is called element by element in the buildr and solver as i recall. take care not to call it twice

philbucher commented:
> this is called element by element in the buildr and solver as i recall. take care not to call it twice

ok but isn't this inconsistent then?
Should we move those calls to the Scheme?

RiccardoRossi commented:
I did it this way to allow for a usage of the type

~~~cpp
for(auto elem: model_part.Elements)
{
      elem.InitializeNonLinearIteration() //eventually here some memory could be allocated

      pScheme->CalculateSystemContributions(*(it.base()), LHS_Contribution, RHS_Contribution, EquationId, CurrentProcessInfo);

        Assemble(A, b, LHS_Contribution, RHS_Contribution, EquationId);

        pScheme->CleanMemory(*(it.base()));
}
~~~

so that an element could eventually allocate and deallocate something in the clean.
but i just checked the code and it looks like this behaviour is long gone

RiccardoRossi reviewed (changes requested):
i'll remove my block. it is used as you are telling already in the staticscheme so the behaviour you are proposing is already standard

philbucher commented:
@RiccardoRossi you are still blocking ...?

RiccardoRossi commented:
yes i am. as far as i see we don t have a solution, and i rrally don t want a half backed kack to.go.official

philbucher commented:
ok jut wanted to clarify

RiccardoRossi commented:
sorry, i thought it is as a different PR. i can definitely remove my block on this

philbucher committed:
Merge branch 'master' into core/scheme-init-nl-iter

philbucher committed:
deprecating old methods

philbucher committed:
missing const

philbucher committed:
remove duplicated calls from BlockBuilder

philbucher committed:
remove duplicated calls from Elim B&S

philbucher committed:
removing duplicated calls from other B&Ss

philbucher committed:
this is now in the baseclass

philbucher committed:
this is now done in baseclass

philbucher committed:
unnecessary

philbucher committed:
done somewhere else

philbucher committed:
now done in BaseClass

philbucher commented:
@KratosMultiphysics/technical-committee in the scope of updating the `Scheme`s I took another look at this and we have to discuss the following things:
- **IN MASTER**: The `FinalizeSolutionStep` is called TWICE on Constraints, in the `Scheme` and in the `BuilderAndSolver`s!
=> I moved it to the `Scheme` such that it is consistent with `Element`s and `Condition`s
- The behavior of calling 
    - `InitializeSolutionStep`
    -  `FinalizeSolutionStep`
    - `InitializeNonLinIteration`
    - `FinalizeNonLinIteration`
   of `Element`s, `Condition`s and `Constraint`s was unified and is done in the `Scheme` (called from the `Strategy` => there is quite a mess going on in some derived `Scheme`s, hence I think this is really necessary
- What [@RiccardoRossi wrote above](https://github.com/KratosMultiphysics/Kratos/pull/5121#issuecomment-531109262) was still done in old `BuilderAndSolver`s, I removed those (since it is now done in the `Scheme`) and deprecated the corresponding functions of the `Scheme`
- I could not find where the `Constraint`s are initialized... This is neither done in the `Scheme` nor in the `BuilderAndSolver` not in the `Strategy`. @adityaghantasala @loumalouomega do you know?
- Some implementations that I am unifying here is done in a more or less complete manner in some derived `Scheme`s. I cleaned those right away otherwise I would again lose the overview
- In many `Scheme`s `InitializeNonLinearIteration` is called (form `Element`s and `Condition`s) before calculating system contributions, esp in the Fluid. Now that this is done in the BaseClass I think/hope that this can be safely removed. The `Scheme`s that are used for the structural problems already do this.

**SORRY THIS IS SUCH A BIG PR!** I would be happy to make smaller PRs! Here I wanted to collect all changes to make sure I don't miss anything!
I would very much appreciate if someone else could take a look through the code (not possible in Github) and check my implementation!

philbucher committed:
unnecessary

philbucher commented:
I updated the Core Schemes and the ones in StructuralMechanics. I took a look at the ones in FluidDynamics, they are a HUGE mess :sweat_smile: 
=> TBD!

philbucher committed:
updated Fluid Schemes
Not sure if it still works

philbucher committed:
missing calls to BaseClass

philbucher committed:
fixing var names

philbucher committed:
Updating Rans App

philbucher reviewed (commented):
I just added the calls for the sake of completeness, I am not sure about the order!
=> Refers to the fluidschemes: https://github.com/KratosMultiphysics/Kratos/pull/5121/commits/6abf352a1ae663be9c72910f71cf97f651e68ce0

philbucher committed:
tmp run nightly tests to get more tests coverage

philbucher committed:
Revert "tmp run nightly tests to get more tests coverage"

This reverts commit ad4b25d60b5b55575d891968a04ba5abe77afec4.

philbucher commented:
ok nightly test also pass

philbucher committed:
Merge remote-tracking branch 'origin/master' into core/scheme-init-nl-iter

philbucher committed:
fixed merge conflicts

philbucher committed:
missing changes in FluidDynamics

philbucher committed:
minor missing

loumalouomega reviewed (approved):
OK (if it passes the tests)

loumalouomega reviewed (changes requested):
Does not pass the tests

rubenzorrilla committed:
Fix structure relaxation scheme

loumalouomega reviewed (dismissed):
Now

rubenzorrilla reviewed (dismissed):
I carefully re-reviewed the files one-by-one and I think this can go in. However, I'm afraid that if any one has added a new scheme by copying & pasting and old one the missuses will still be there. Maybe it is a good idea that we merge the master and grep again for `InitializeNonLinearIteration`.

philbucher committed:
Merge branch 'master' into core/scheme-init-nl-iter

philbucher committed:
using EntitiesUtilities

philbucher committed:
corrections in RANS

philbucher committed:
typo

philbucher committed:
leftover

philbucher committed:
final changes

philbucher committed:
missing core update

philbucher committed:
removed one line too many

rubenzorrilla reviewed (approved):
I think that there are no more leftovers.
